docs/performance/inserts/parallel.txt

```diff
@@ -0,0 +1,88 @@
+.. _concurrent_inserts:
+
+================
+Parallel Inserts
+================
+
+If you have a queue of inserts, one way to process them in your client is to
+send them one by one, waiting for the response from previous insert before
+sending the next one.
+
+With a latency of 5ms between the server and client, even if inserts happened
+instantaneously, this client could only ever do 200 inserts per second.
+
+If you're handling a lot of inserts, this sort of setup can very quickly become
+a performance bottleneck.
+
+The solution to this is to send multiple inserts concurrently. That is, send
+off every insert request as soon as you need to and do not wait for a
+response before sending another insert.
+
+This document introduces you to a number of different approaches for doing
+concurrent inserts.
+
+.. rubric:: Table of Contents
+
+.. contents::
+   :local:
+
+Example
+=======
+Suppose we have a stream of data we want to persist into CrateDB.
+Parallelising this processing in java is possible using ``CompletableFuture``: 
+
+```
+IntStream.iterate(0, i -> i + 2)
+    .limit(1000)
+    .forEach(i -> {
+        CompletableFuture<Integer> insertFuture =
+            CompletableFuture.supplyAsync(() -> {
+                try {
+                    PreparedStatement stmt =
+                      connection.prepareStatement("insert into t1 values (?)");
+                    stmt.setInt(1, i);
+                    return stmt.executeUpdate();
+                } catch (SQLException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+
+        insertFuture.whenComplete((Integer result, Throwable failure) -> {
+            if (failure == null) {
+                // use row count
+            } else {
+                // handle insert failure
+            }
+        });
+    });
+```
+This way the inserts will be supplied to be executed asynchronously on the 
+``commmonPool`` (you can also provide your own ``Executor`` using the 
+appropriate ``supplyAsync`` signature).
```

Comment:
Maybe we should also add a note to clarify that this if possible, it would also greatly improve performance if the inserts were batched and cross-link to the other documents? (it would actually the first thing that should be done if possible - before trying to add parallelism)

Reply:
I agree. I think that should be part of the introduction though, saying that explicit parallelism should be explored in case batching writes doesn't already provide the desired performance or something along those lines.